apollo-router/src/drop_watch.rs

```diff
@@ -0,0 +1,141 @@
+//! Provide a [`DropWatch`] utility.
+//!
+//! Accept a FnOnce and carefully watch over it until we are dropped.
+//!
+//! Useful when we want to join threads which will ultimately be terminated
+//! within an asynchronous context.
+
+use std::mem::ManuallyDrop;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+pub(crate) struct DropWatch {
+    watcher_handle: ManuallyDrop<JoinHandle<()>>,
+    park_flag: Arc<AtomicBool>,
+}
+
+impl DropWatch {
+    /// Create a DropWatch which will execute the supplied closure then wait for notification
+    pub(crate) fn run_and_wait<F: FnOnce() + Send + 'static>(watched: F) -> Self {
+        let park_flag = Arc::new(AtomicBool::new(false));
+        let watching_flag = park_flag.clone();
+        let watcher_handle = std::thread::spawn(move || {
+            watched();
+            // Park the thread until this instance is dropped (see Drop impl)
+            // We may actually unpark() before this code executes or exit from park() spuriously.
+            // Use the watching_flag to control a loop which waits for the flag to be updated
+            // from Drop.
+            while !watching_flag.load(Ordering::Acquire) {
+                std::thread::park();
+            }
+        });
+        Self {
+            park_flag,
+            watcher_handle: ManuallyDrop::new(watcher_handle),
+        }
+    }
+
+    /// Create a DropWatch which will wait for notification before executing the supplied closure
+    pub(crate) fn wait_and_run<F: FnOnce() + Send + 'static>(watched: F) -> Self {
+        let park_flag = Arc::new(AtomicBool::new(false));
+        let watching_flag = park_flag.clone();
+        let watcher_handle = std::thread::spawn(move || {
+            // Park the thread until this instance is dropped (see Drop impl)
+            // We may actually unpark() before this code executes or exit from park() spuriously.
+            // Use the watching_flag to control a loop which waits for the flag to be updated
+            // from Drop.
+            while !watching_flag.load(Ordering::Acquire) {
+                std::thread::park();
+            }
+            watched();
+        });
+        Self {
+            park_flag,
+            watcher_handle: ManuallyDrop::new(watcher_handle),
+        }
+    }
+}
+
+impl Drop for DropWatch {
+    fn drop(&mut self) {
+        // Safety: watcher_handle is taken, and must not be accessed again
+        // Since this is in Drop::drop, it is not possible to access it afterwards.
+        // https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html#method.into_inner
+        let wh = unsafe { ManuallyDrop::<JoinHandle<()>>::take(&mut self.watcher_handle) };
+        self.park_flag.store(true, Ordering::Release);
+        wh.thread().unpark();
+        wh.join().expect("drop watcher thread terminating");
```

Comment:
I am not sure this is the behaviour we want. We need the various threads to be joined before stopping the process, but here that would be at the cost of an executor thread blocked waiting for the `DropWatch` to execute its stack. Which means that every time we reload schema or config, dropping the old tracer provider and meter provider would affect the async runtime

Reply:
It's hard to say which is worse: spawning unbound tasks/threads (complicated debugging) or possible synchronisation delays (are these going to be long in duration?). I think the former, but I can see there's a case for the latter and we should benchmark out the impact.

If we are trying to address the former, then the runtime would be affected however we do it. The constraint is that we wish to "join" (either `join` for sync context or `await` for async context) a spawned `thread`/`task`. Right now we don't do that and it is a source of debugging confusion and (potentially) a source of future bugs.

The main problem is trying to execute `sync` code from a `sync` fn which "lives within" an `async` call stack. We can't change the behaviour of `Drop`, so we need a way to be able to address the problem. This is the least intrusive mechanism I can concoct.

You could try to concoct something which enters the async runtime during `Drop`, but that might be executing after tokio has finished and you still won't have a way to `await` the spawned task. (You could work around the "runtime has finished" problem using `try_current()` but you still wouldn't have synchronisation point if there was a runtime.

Reply:
the main goal of the code before was to execute that synchronous task outside of the async executor. So joining on that thread in the async executor is a regression

Reply:
My point still stands. The existing code (telemetry) just spawned a thread and didn't join it in. The new code does the same thing but joins. We could take out the join, but then we'd be back where we were before but with a slightly tidier utility class.

Reply:
To rephrase my concerns here:
- dropping tracer providers in the async code was causing issues because it is a potentially long synchronous task, so it was moved to a thread
- I am not attached to the thread solution, and moving that pattern in a specific structure that encodes a correct behaviour sounds fine
- reintroducing the synchronous task in the async runtime is not a correct behaviour
- I agree that we need a way to track the threads we start, and we have to join on them at some point
- when to join on them is unclear to me at this point. Do you want to make sure that all threads have stopped before closing the process? Or is it during before switching to a new conf during a reload?
- as a side point, it is unclear to me how threads cause issues in tracking memory leaks with heaptrack, even with frequent reloads, but we can look into that elsewhere